packages/loader/container-loader/src/connectionHandler.ts

```diff
@@ -0,0 +1,886 @@
+
+    /**
+     * Tells if container is in read-only mode.
+     * Data stores should listen for "readonly" notifications and disallow user
+     * making changes to data stores.
+     * Readonly state can be because of no storage write permission,
+     * or due to host forcing readonly mode for container.
+     * It is undefined if we have not yet established websocket connection
+     * and do not know if user has write access to a file.
+     */
+    private get readonly() {
+        if (this._forceReadonly) {
+            return true;
+        }
+        return this._readonlyPermissions;
+    }
+
+    public get readOnlyInfo(): ReadOnlyInfo {
+        const storageOnly = this.connection !== undefined && this.connection instanceof NoDeltaStream;
+        if (storageOnly || this._forceReadonly || this._readonlyPermissions === true) {
+            return {
+                readonly: true,
+                forced: this._forceReadonly,
+                permissions: this._readonlyPermissions,
+                storageOnly,
+            };
+        }
+
+        return { readonly: this._readonlyPermissions };
+    }
+
+    private static detailsFromConnection(connection: IDocumentDeltaConnection): IConnectionDetails {
+        return {
+            claims: connection.claims,
+            clientId: connection.clientId,
+            existing: connection.existing,
+            checkpointSequenceNumber: connection.checkpointSequenceNumber,
+            get initialClients() { return connection.initialClients; },
+            maxMessageSize: connection.serviceConfiguration.maxMessageSize,
+            mode: connection.mode,
+            serviceConfiguration: connection.serviceConfiguration,
+            version: connection.version,
+        };
+    }
+
+    constructor(
+        private readonly serviceProvider: () => IDocumentService | undefined,
+        private client: IClient,
+        reconnectAllowed: boolean,
+        private readonly logger: ITelemetryLogger,
+        private readonly props: IConnectionManagerFactoryArgs,
+    ) {
+        this.clientDetails = this.client.details;
+        this.defaultReconnectionMode = this.client.mode;
+        this._reconnectMode = reconnectAllowed ? ReconnectMode.Enabled : ReconnectMode.Never;
+
+        // Outbound message queue. The outbound queue is represented as a queue of an array of ops. Ops contained
+        // within an array *must* fit within the maxMessageSize and are guaranteed to be ordered sequentially.
+        this._outbound = new DeltaQueue<IDocumentMessage[]>(
+            (messages) => {
+                if (this.connection === undefined) {
+                    throw new Error("Attempted to submit an outbound message without connection");
+                }
+                this.connection.submit(messages);
+            });
+
+        this._outbound.on("error", (error) => {
+            this.props.closeHandler(normalizeError(error));
+        });
+    }
+
+    public dispose(error: any) {
```

Comment:
This seems idempotent but I'm used to seeing an `if (this.closed) { return; }` block at the top, as a defensive measure.